runtime/java/src/common/RandomGen.java

```diff
@@ -0,0 +1,42 @@
+package common;
+
+import java.util.*;
+
+import common.exceptions.*;
+import silver.core.*;
+
+public final class RandomGen {
+	public static Object runRandomGen(final OriginContext ctx, final int seed, final NRandomGen r) {
+		return runRandomGen(ctx, new Random(seed), r);
+	}
+
+	public static Object runRandomGen(final OriginContext ctx, final NRandomGen r) {
+		return runRandomGen(ctx, new Random(), r);
+	}
+
+	// "Interpret" a RandomGen computation using the specified Random generator
+	@SuppressWarnings("unchecked")
+	public static Object runRandomGen(final OriginContext ctx, final Random rng, final NRandomGen r) {
+		if (r instanceof PmapRandomGen) {
```

Comment:
should this check the forward?

Reply:
That would introduce a fair bit of overhead, since we would need to decorate the `RandomGen` into a tree... and I can't think of how that could possibly be useful, anyway.  So I'm just making it throw an error if we ever get a production that we didn't expect.  

... I guess this could be another argument in favor of having data types in Silver.  I'm still not convinced though :-P